### PR TITLE
Tweak: Add duplicate post rest for pages panel [ED-10867]

### PR DESCRIPTION
**Pull request by lenalElementor**

### CODE REVIEW

File: modules/site-navigation/data/endpoints/duplicate-post.php [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+	public function get_format() {
+		return 'site-navigation/duplicate-post';
+	}
+
+	public function create_items( $request ) {
+		$post_id = $request->get_param( 'post_id' );
+		$post_title = $request->get_param( 'title' );
+
+		$post = get_post( $post_id );
```
Comment (ranshoshani34) [inline modules/site-navigation/data/endpoints/duplicate-post.php:50]: Please handle `if (!$post)`

File: modules/site-navigation/data/endpoints/duplicate-post.php [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+		$new_post_id = $this->duplicate_post( $post, $post_title );
+
+		if ( is_wp_error( $new_post_id ) ) {
+			return new \WP_Error( 500, sprintf( 'Error while duplicating post.' ) );
```
Comment (ranshoshani34) [inline modules/site-navigation/data/endpoints/duplicate-post.php:55]: why `sprintf`?

File: modules/site-navigation/data/endpoints/duplicate-post.php [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+	/**
+	 * Duplicate post
+	 *
+	 * @param $post
+	 *
+	 * @return int|\WP_Error
+	 */
+	private function duplicate_post( $post, $post_title ) {
+		$page_status = 'draft';
```
Comment (ranshoshani34) [inline modules/site-navigation/data/endpoints/duplicate-post.php:77]:
```suggestion
		$post_status = 'draft';
```

File: modules/site-navigation/data/endpoints/duplicate-post.php [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+			return;
+		}
+
+		foreach ( $post_meta as $key => $values ) {
+			if ( '_wp_old_slug' === $key ) { // Ignore this meta key
+				continue;
+			}
+
+			foreach ( $values as $value ) {
+				$value = maybe_unserialize( $value );
```
Comment (ranshoshani34) [inline modules/site-navigation/data/endpoints/duplicate-post.php:119]: why do we need to unserialize and slash?

File: modules/site-navigation/data/endpoints/duplicate-post.php [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+	}
+
+	/**
+	 * duplicate_post_taxonomies
+	 *
+	 * @param int $post_id
+	 * @param int $new_post_id
+	 */
+	private function duplicate_post_taxonomies( $post_id, $new_post_id ) {
+		$taxonomies = array_map( 'sanitize_text_field', get_object_taxonomies( get_post_type( $post_id ) ) );
```
Comment (ranshoshani34) [inline modules/site-navigation/data/endpoints/duplicate-post.php:132]: why sanitize?

File: modules/site-navigation/data/endpoints/duplicate-post.php [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+		}
+
+		foreach ( $taxonomies as $taxonomy ) {
+			$post_terms = wp_get_object_terms( $post_id, $taxonomy, [ 'fields' => 'slugs' ] );
+
+			if ( is_wp_error( $post_terms ) ) {
+				continue;
+			}
+
+			wp_set_object_terms( $new_post_id, $post_terms, $taxonomy, false );
```
Comment (ranshoshani34) [inline modules/site-navigation/data/endpoints/duplicate-post.php:145]:
```suggestion
			if ( ! is_wp_error( $post_terms ) ) {
			    wp_set_object_terms( $new_post_id, $post_terms, $taxonomy, false );
			}
```

I think it's more readable

File: tests/phpunit/elementor/modules/site-navigation/data/recent/test-controller.php [hunk 7]
```diff
@@ -233,6 +234,49 @@ public function test_create_items() {
...
+			'post_id' => 'test',
+		];
+		$response = $this->send_request( 'POST', self::DUPLICATE_POST_ENDPOINT, $params );
+		// Assert.
+		$this->assertEquals( 400, $response->get_status() );
+	}
+
+	public function test_create_items_duplicate_post() {
+		// Arrange.
+		$this->act_as_editor();
```
Comment (ranshoshani34) [inline tests/phpunit/elementor/modules/site-navigation/data/recent/test-controller.php:256]: Can you add test for a role that cant create posts?

**Review by ranshoshani34 — CHANGES_REQUESTED** [review]
Good job. Please see requested changes :)

File: modules/site-navigation/data/endpoints/duplicate-post.php [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+	}
+
+	public function create_items( $request ) {
+		$post_id = $request->get_param( 'post_id' );
+		$post_title = $request->get_param( 'title' );
+
+		$post = get_post( $post_id );
+
+		if ( ! $post ) {
+			return new \WP_Error( 500, printf( 'Error while duplicating post.' ) );
```
Comment (ranshoshani34) [inline modules/site-navigation/data/endpoints/duplicate-post.php:53]:
```suggestion
			return new \WP_Error( 500, 'Post not found'  );
```

File: modules/site-navigation/data/endpoints/duplicate-post.php [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+		}
+
+		$new_post_id = $this->duplicate_post( $post, $post_title );
+
+		if ( is_wp_error( $new_post_id ) ) {
+			return new \WP_Error( 500, printf( 'Error while duplicating post.' ) );
```
Comment (ranshoshani34) [inline modules/site-navigation/data/endpoints/duplicate-post.php:59]:
```suggestion
			return new \WP_Error( 500, 'Error while duplicating post.'  );
```

What i meant

File: modules/site-navigation/data/endpoints/duplicate-post.php [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+	 * Duplicate post
+	 *
+	 * @param $post
+	 *
+	 * @return int|\WP_Error
+	 */
+	private function duplicate_post( $post, $post_title ) {
+		$post_status = 'draft';
+		$current_user = wp_get_current_user();
+		$new_page_author = $current_user->ID;
```
Comment (ranshoshani34) [inline modules/site-navigation/data/endpoints/duplicate-post.php:83]:
```suggestion
		$new_post_author = $current_user->ID;
```

😅

File: tests/phpunit/elementor/modules/site-navigation/data/recent/test-controller.php [hunk 11]
```diff
@@ -233,6 +234,67 @@ public function test_create_items() {
...
+		$this->assertTrue( $duplicated_post_id > $post );
+	}
+	
+	public function test_create_items_duplicate_post__forbidden() {
+		// Arrange.
+		$this->act_as_subscriber();
+
+		// Act.
+		$params = [
+			'post_id' => 'test',
```
Comment (ranshoshani34) [inline tests/phpunit/elementor/modules/site-navigation/data/recent/test-controller.php:287]: send a real id to mock the right conditions

**Review by ranshoshani34 — CHANGES_REQUESTED** [review]

File: tests/phpunit/elementor/modules/site-navigation/data/recent/test-controller.php (hunk 11 above)
Comment (lenalElementor) [reply tests/phpunit/elementor/modules/site-navigation/data/recent/test-controller.php:287]: This test checks wrong post id, so string is sent...

Comment (ranshoshani34) [reply tests/phpunit/elementor/modules/site-navigation/data/recent/test-controller.php:287]: This is the test for forbidden role

**Review by ranshoshani34 — CHANGES_REQUESTED** [review]

**Review by ranshoshani34 — APPROVED** [review]

**Pull request merged**